### ISSUE TITLE
BlenderBIM feature request: Additional properties within Parametric Door

### ISSUE THREAD

**Issue by trhyder** [IfcOpenShell/IfcOpenShell#3537]
Below are some additional features that would allow the Parametric Door tool to achieve more use cases and improve the user experience.

1. Swing operation should ideally be controlled from the object (not the type). This would reduce the number of door types by up to half.
> Edit: This would require an extension to the official properties. IfcDoor has an ObjectPlacement parameter but the documentation says that it is only for the y-axis (in and out). Strangely though, the ifc4.3 docs say that an IfcDoor can store the width and height, which currently BBim doesn't seem to do. I wonder how BBim recognizes doors that have had their height stored this way as BBim stores it in ifcDoorType.

2. An option to tell the door type to size the Lining Thickness to equal the Wall Thickness dynamically. Depending on the design intent this would also reduce the number of door types on large projects considerably.
> Edit: This should be supported by BBim as the ifc4.3 docs says: "LiningDepth, if omitted, equal to wall thickness - this only takes effect if a value for LiningThickness is given. If both parameters are not given, then there is no lining." But currently, the BBim interface does not allow the user to omit a value (by deleting it or by setting the value to zero).

3. Better clarity on the Lining to Panel Offset X and Y properties. I think these properties might be trying to achieve different lining (frame) profile types, but if they are, they often produce overlapping geometry. There are 3 different types of profiles, Block, Simple rebate, and Double rebate which would probably cover 95% of use cases. The parameters that could achieve these 3 types are Panel Rebate Thickness (instead of Panel Offset X), Panel Rebate Depth (instead of Panel Offset Y), and a boolean toggle for Double Rebate (which would also require an additional new property value for the Panel Stop Depth)
![DOOR FRAME DETAIL](https://github.com/IfcOpenShell/IfcOpenShell/assets/59872808/c4477468-1002-4139-868c-04c5e755bddd)
> Edit: Block and Simple rebate frames are achievable by BBim. Double rebate frames are not supported by the ifc4.3 docs so this would be an extension to the official properties. The labeling of the properties does follow the ifc4.3 docs which are unfortunately not obvious on first-time use. I also got a little confused by not knowing that the geometry doesn't fully update until you click Finish Editing. So my comment about overlapping geometry is not correct. I am not sure how this will work if the Lining Depth is derived from the wall thickness. I suspect not as desired because the offset value is from the Outer side face.

4. Support for a few new Door Operation Types:
  * ~~EMPTY_OPENING~~
  * ~~SINGLE_FOLDING~~
  * ~~DOUBLE_FOLDING~~
  * DOUBLE_ACTING
  * FOLDING
  * ROLLINGUP
  * ~~SINGLE_PANEL_LIFT~~
  * REVOLVING
  * FIXEDPANEL
> Edit: I adjusted the above list to show the [Type values](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcDoorPanelOperationEnum.htm#7.1.2.1.2-Type-values) that ifc4.3 document 

5. The ability to toggle on/off door components within the door type parameters, such as:
 * Lining (frame) - currently not possible
  > Edit: This should be supported as the ifc4.3 docs says: "LiningDepth, if omitted, equal to wall thickness - this only takes effect if a value for LiningThickness is given. If both parameters are not given, then there is no lining."

 *  Panel - currently not possible
  > Edit: The ifc4.3 docs describe this an an optional property. Currently in BBim, if the PanelDepth is set to 0 an error is returned on Finish Editing. I would expect a zero PanelDepth to show no panel.

 *  Casing - currently possible by setting the Casing Thickness to zero
  > Edit: Setting the Casing Depth in BBim to zero should also disable casings as the ifc4.3 docs list these both as optional properties. Currently, setting the Casing Depth to 0 in BBim causes an error on Finish Editing.

 *  Threshold - currently possible by setting the Threshold Thickness to zero

 *  Handle - currently not possible
  > Edit: I can't find this in the ifc file or ifc 4.3 docs, is BBim hardcoded door handles?
These options would provide more detailed functionality and varied design options.

6. Ability to control the inside and outside casings separately. This would allow for doors in walls that have non-identical faces. Also, a parameter to control the position of the casing would provide more flexibility (Casing Offset). My personal preference would be for casings to be switched off by default.
> Edit: This would need to be an extension to the official properties

7. A select option to set the threshold to linings or casings. Also depending on the design intent this would reduce the number of door types. My personal preference would be for the threshold to be switched off by default.
> Edit: Not supported in the ifc4.3 docs as I describe above. However, the docs do say "If omitted (and with a given value to threshold thickness) it indicates an adjustable depth (i.e. a depth that adjusts to the thickness of the wall into which the occurrence of this door style is inserted)" which BBim currently does not follow.

8. It would be a helpful design aid to see other door size properties in addition to Overall Height/Width like:
* Door Panel
> Edit: Possible to show. Panel Width = (Overall - 2*CasingThickness + 2*LiningToPanelOffsetX) Panel Height = (Overall - CasingThickness + LiningToPanelOffsetX - Threshold Thickness)
* Rough Opening (would also require an additional parameter for the gap)
> Edit: This would require an extension to the official properties
* Egress Clearance
> Edit: Possible to show. Egress Width = (Overall - 2*CasingThickness) Egress Height = (Overall - CasingThickness - Threshold Thickness)
* Framed Unit (Overall less casings)
> Edit: Possible to show. Unit Width  = (Overall - 2*CasingThickness + 2*LiningThickness) Unit Height = (Overall - CasingThickness + LiningThickness - Threshold Thickness)

**Comment by trhyder** [comment]
Thanks for pointing me to the docs, I am happy to explore this further and will have a deeper look. My first observation is that some of the details used to describe the parameters appear to contradict the definition.

The definition says:

> The door lining is the frame which enables the door leaf to be fixed in position. The door lining is used to hang the door leaf. 

But from my experience, it would be near impossible to hang a door panel in front of the lining (frame), as this would cause the door panel to bind at the hinge end. Also, a door panel should be installed between the linings (frames) so that it can latch into the frame.
<img width="1031" alt="Screenshot 2023-08-03 at 10 28 49 AM" src="https://github.com/IfcOpenShell/IfcOpenShell/assets/59872808/ee6506d8-58d0-41d5-9c08-d83eb5c67a46">

But if this is the IFC way, I feel like I must be missing something. I'll look more into it and report back.

**Comment by trhyder** [comment]
I edited my original post above with my progress. Input and feedback appreciated!

**Comment by trhyder** [comment]
@Moult would you prefer I separated these into individual issues, or is it best all in one?

**Comment by Moult** [comment]
Perhaps best to leave it here until I find the time to digest it. I generally agree, but there are nuances and things we've chosen to do that make sense for spec compliance but may not make sense to users that perhaps we need to rethink. When I find the time, would you be up for a call to talk over this issue in detail?

**Comment by trhyder** [comment]
Sounds good. Let me know when you have a moment.

**Comment by trhyder** [comment]
Hey @Moult, I am keen to put some hours into moving this item forward. I think with an initial discussion I could then do some more research to report back on a proposal. I also have some coding skills and might be able to implement what is agreed. Let me know when you have some time.

**Comment by trhyder** [comment]
@Moult, circling back on this one a year later.

One thing that still stands out is that the Lining Depth of a door frame can't match the the thickness of the wall it is placed in. When I delete the Lining Depth value it reverts back to a default 0.05m, and if I enter 0 then I get an error.

Should Bonsai support Lining Depth = wall thickness?

**Comment by Moult** [comment]
I think the best way forward is to implement a new type of parametric door, that is consciously non compliant with the current standard case door but also much more practical (I.e. handles no, single, and double rebate properly and overcomes shortcomings like door hang direction or explicit lining thickness).

I see two issues here: 1) the standard case parametric door has shortcomings. Let's solve them and propose a new standard. 2) door geometry is generated at the type level, whereas we need occurrence spevifkc

**Comment by trhyder** [comment]
Excellent direction - I'm excited. I’ll start working on a draft proposal. I’ve recently had an intern cad up our standard door details, so I’m in the zone.

I have an initial question: Are you open to representing the frame as an extruded profile instead of using multiple parameters? From my perspective, the key benefit would not necessarily be about the level of detail shown in the model but rather the ability to share jamb types across multiple door types.

For reference, I'm adding a link to the issue I created for Ifc4.3 https://github.com/buildingSMART/IFC4.3.x-development/issues/673

Additionally, would you like me to create a separate issue regarding the generation of door geometry at the occurrence level rather than at the type level, or is that already in the pipeline?

**Comment by Moult** [comment]
Yes I think you're really the best person for this task. I'd like to reiterate that for a first version, do not allow the ability to nominate an extruded profile. Constrain yourself to just using parameters. This is for simplicity of implementation (swept profiles can be hard). Later on, a more advanced door generator can be created.

I'm available anytime to discuss / finalise the parameters (I assume you're going to continue where you left off [here](https://github.com/buildingSMART/IFC4.3.x-development/issues/673#issuecomment-1837570985)) anytime :)

**Comment by Moult** [comment]
https://community.osarch.org/discussion/2777/designing-a-new-window-door-generator/
